### PR TITLE
fix: publishing config to respect `enableHierarchyDisplay`

### PR DESCRIPTION
**Pull request by Harshita-mindfire**
Related issue: https://github.com/dendronhq/dendron/issues/2943
Also resolves these enhancements: https://github.com/dendronhq/dendron/issues/581 https://github.com/dendronhq/dendron/issues/2824

Co-authored by : kevin
# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
NA

### Extended
NA
### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [ ] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [x] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
NA

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

File: packages/engine-server/src/markdown/remark/hierarchies.ts [hunk 1]
```diff
@@ -65,15 +65,22 @@ function footnoteDef2html(definition: FootnoteDefinition) {
...
+  let hierarchyDisplay = config?.hierarchyDisplay;
+
+  if (MDUtilsV5.shouldApplyPublishingRules(proc)) {
+    hierarchyDisplay =
+      ConfigUtils.getHierarchyDisplayConfigForPublishing(
+        config
+      ).hierarchyDisplay;
+    hierarchyDisplayTitle =
+      ConfigUtils.getHierarchyDisplayConfigForPublishing(config)
+        .hierarchyDisplayTitle || "Children";
```
Comment (SeriousBug) [inline packages/engine-server/src/markdown/remark/hierarchies.ts:81]: This is a little confusing, does `config` we get from `getProcData` meant to override what's set in the configuration, or does the publishing config override whatever is passed in? Why do we have the `|| "Children"` fallback twice? Making this clearer might be worth it even if it's a bit more verbose.

**Review by SeriousBug — COMMENTED** [review]
This is supposed to disable the display of `Children: ...` and `Backlinks: ...` at the end of published notes right? I tried it out in the test-workspace by exporting it and using the nextjs-template, but it doesn't seem to be disabling it for me. Maybe I'm misunderstanding it or doing something wrong, can you confirm with manual testing that this works in publishing?

**Comment by Harshita-mindfire** [comment]
> This is supposed to disable the display of `Children: ...` and `Backlinks: ...` at the end of published notes right? I tried it out in the test-workspace by exporting it and using the nextjs-template, but it doesn't seem to be disabling it for me. Maybe I'm misunderstanding it or doing something wrong, can you confirm with manual testing that this works in publishing?

The `enableHierarchyDisplay` is responsible for displaying `Children` on the published page. This config does not affect the `Backlinks`. Updated the test-workspace/dendron.yml with `enableHierarchyDispplay: false` and `hierarchyDisplayTitle: Sub Notes` setting. If you do `dendron publish init` and then `dendron publish dev`, the `Children` should not be visible when `enableHierarchyDisplay` is set to false.

File: packages/engine-server/src/markdown/remark/hierarchies.ts (hunk 1 above)
Comment (Harshita-mindfire) [reply packages/engine-server/src/markdown/remark/hierarchies.ts]: the config we get from `getProcData` could be v4 or v5, for publishing hierarchies, we get the hierarchy display configs from utility `getHierarchyDisplayConfigForPublishing` that gives us the values from appropriate version. 
Updated the Children fallback twice mistake.

**Review by kevinslin — APPROVED** [review]

**Comment by kevinslin** [comment]
approved, can you merge according to [[Async Update to Pr Merge Process|dendron://private/user.kevin.journal.2022.05.20.m.async-update-to-pr-merge-process]]

**Pull request merged**